### DISCUSSION TITLE
Integration with web components

### DISCUSSION THREAD

**Discussion by hypeJunction** [chakra-ui/zag#1210] [General]
It's getting tiring, building and rebuilding things in React and Vue, so I am wondering if we should take the plunge and go in the direction of web components. I really like what zagjs is doing, as it seems, just like me, you are fed up with rewriting the same code over and over from project to project. But before I do any further digging, what's your take on making zagjs working with web components, anything you see that might make it difficult or impossible? I was thinking lit or webc integration could be an interesting thing to try out.

**Comment by segunadebayo** [comment]
Totally feel you @hypeJunction

We’re also all into modeling components once and reusing everywhere possible.

Regarding web components, we’re open to seeing a Lit implementation similar to Shoelace https://github.com/shoelace-style/shoelace, or Stencil.

Feel free to open a draft PoC of how this could work.

**Reply by max-scopp** [reply]
I've tried a thing with stencil: https://github.com/max-scopp/zag-stencil-demo/blob/master/src/components/my-component/my-component.tsx#L52-L83

I dislike how I re-instatiate the VanillaMachine every time the component's props change, just so when the component is changed externally the machine is kept up to date.

LMKWYT

**Comment by hypeJunction** [comment]
Just doing some legwork reading. This could be a showstopper https://github.com/lit/lit/issues/923

**Reply by segunadebayo** [reply]
There’s an helper library that can help with this
https://github.com/open-wc/open-wc/blob/master/packages/lit-helpers/CHANGELOG.md

**Comment by hypeJunction** [comment]
Perhaps the connector could expose a render function that spells out the template with all the props and events, but it might require quite some work to express all component machines in that manner.

**Comment by gperdomor** [comment]
Hi folks... I added Lit examples again :D
This is the PR https://github.com/chakra-ui/zag/pull/2608

The focus of the MR is to provide a basic integration between Zag and Lit...

A few highlights:
- The base Component class already handle the id creation
- There is no props in the example components, just Zag integration, this means that you can't change the image in the Avatar component or pass custom items to combobox
- Combobox have issues with item list, are highlighted with blue color but not return to original
- spreadProps function was removed in favor of spread from `@open-wc/lit-helpers` package
- normalizeProps was updated to prefix boolean keys with `?` and replace `on` by `@` in event handlers

**Reply by tala91** [reply]
Hi @gperdomor! 👋

Great work on this Lit integration! I've been testing your implementation because we're also very interested in using Zag to build a web components library with Lit.

I can confirm the blue highlighting issue you mentioned with the combobox, and I also found that **the popover positioning was broken** (always appearing in top-left corner instead of positioning correctly relative to the trigger).

However, I found a **solution for the positioning issue**! 🎉

Adding this to the `initMachine` call fixes the popover positioning:

getRootNode: () => this.renderRoot,

After adding this, the popover/dropdown correctly appends to the shadow root and positions properly relative to its trigger element. This makes sense because Lit uses shadow DOM by default, and Zag needs to know the correct root node context for proper DOM manipulation and positioning calculations.

This might be worth adding to your base Component class or Reactive Controller implementation to ensure all Zag components work correctly with Lit's shadow DOM.

Thanks for pioneering this integration - Popover it's working great now with this fix! Would be fantastic to see this merged. 🚀

**Reply by gperdomor** [reply]
@tala91 thank you for your words 😄... I updated the code with your suggestions...

**Comment by gperdomor** [comment]
One more thing... I added the base `Component` class to the examples, but I think this is not the best approach, probably a Reactive controller is better for this... "A reactive controller is a separate object owned by a component"... In the same UI library I can have normal components without zag machine and other components which owns or contains an zag machine both extending the same base class, but the second one using the reactive controller...

**Reply by gperdomor** [reply]
I added a new Zag Reactive Controller and created a new Avatar component using it... Please let me know WDYT @segunadebayo @anubra266 @max-scopp

**Reply by nikparo** [reply]
I'd say a ReactiveController based system is definitely the way to go, and I borrowed from your work heavily for https://github.com/chakra-ui/zag/pull/2698

**Reply by gperdomor** [reply]
@nikparo that is amazing man, im glad to see that my work was used as base for that. I also think that the controller is the way to go considering that maybe not all components in the library needs a machine, like a label component for example…

**Reply by nikparo** [reply]
There are also components that need multiple machines, such as [menu-nested.ts](https://github.com/chakra-ui/zag/pull/2698/files#diff-aec33fdce1690affaa9af79336a95284dc01ff5121ed05f24d8693a05ffaa900)

**Reply by tala91** [reply]
Hi @nikparo,

first of all thanks for your work on the zag/lit wrapper, it’s been really useful!
I’m testing the Switch component and I noticed something strange: when I try to use it in a controlled way, the onCheckedChange callback seems to fire only every other time (once yes, once no).

Have you already experimented with handling the checked state in controlled mode?
Also, regarding default values passed as properties (e.g. defaultChecked), in order to read them correctly I had to move the machine initialization inside the render function and add a condition so that it only runs the first time.

Did you have a different approach in mind for this?

Thanks a lot!

**Reply by nikparo** [reply]
Thanks @tala91! It's an interesting project and happy it's of use, unfortunately I've been a bit strapped on time and haven't made much progress recently.

> when I try to use it in a controlled way, the onCheckedChange callback seems to fire only every other time (once yes, once no).

I haven't experimented much with default values, sorry. If it's Lit specific then my initial guess would be that it has to do with the `normalizeProps` function or Lit property handling. If it is a `defaultValue`, then does it e.g. become `undefined` and get filtered out somewhere along the line, resulting in no change?

> machine initialization inside the render function

I found this to be problematic elsewhere, since some machines as part of their initialization will look for nodes by id. If they haven't been added to the dom yet then the machine can break (e.g. missing event listeners)

**Reply by tala91** [reply]
Hi, I think I’ve figured out the issue with the Switch in controlled mode.
In the bindable.ts file, inside the set function, I added this line:
`store.value = next;`

Here’s the updated snippet:
`

        set(nextValue: T | ((prev: T) => T)) {
            
            const prev = store.value;
            const next = isFunction(nextValue) ? nextValue(prev as T) : nextValue;
            
            if (props().debug) {
                console.log(`[bindable > ${props().debug}] setValue`, { next, prev });
            }
            
            if (!controlled()) store.value = next;
            if (!eq(next, prev)) {
                store.value = next; // Fix for controlled mode
                props().onChange?.(next, prev);
            }
        }
`

With this change, the onCheckedChange callback now fires consistently on every toggle, even in controlled mode.
